File: src/1329. Sort the Matrix Diagonally.cpp
class Solution {
public:
    vector<vector<int>> diagonalSort(vector<vector<int>>& mat) {
        int m = mat.size();
        int n = mat[0].size();
        unordered_map<int, priority_queue<int, vector<int>, greater<int>>> candidates;
        for (int i = 0; i < m; ++i) {
            for (int j = 0; j < n; ++j) {
                int diff = i - j;
                candidates[diff].push(mat[i][j]);
            }
        }
        for (int i = 0; i < m; ++i) {
            for (int j = 0; j < n; ++j) {
                int diff = i - j;
                mat[i][j] = candidates[diff].top();
                candidates[diff].pop();
            }
        }
        return mat;
    }
};

File: src/1329. Sort the Matrix Diagonally.js
/**
 * @param {number[][]} mat
 * @return {number[][]}
 */

export function diagonalSort(mat) {
  const m = mat.length,
    n = mat[0].length,
    candidates = new Map()
  for (let i = 0; i < m; i++) {
    for (let j = 0; j < n; j++) {
      const diff = i - j
      if (!candidates.has(diff)) candidates.set(diff, [])
      candidates.get(diff).push(mat[i][j])
    }
  }
  for (let [diff, diagonalItems] of candidates) {
    diagonalItems.sort((a, b) => b - a)
  }
  for (let i = 0; i < m; i++) {
    for (let j = 0; j < n; j++) {
      const diff = i - j
      mat[i][j] = candidates.get(diff).pop()
    }
  }
  return mat
}
